### PR TITLE
More helper code for nodegroups, eg AtomicInboxBoundaryData, number_of_block_boundaries for an ElementId

### PR DESCRIPTION
**Pull request by nilsdeppe**
### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

**Review by wthrowe — DISMISSED** [review]

File: src/Evolution/DiscontinuousGalerkin/AtomicInboxBoundaryData.hpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+   * `OrientationMap` to be put in the same block frame as the element that is
+   * receiving the data (i.e. that whose inbox this is being inserted into).
+   *
+   * The hash is computed as
+   * \f{align}{
+   * 2^D d + 2^{D-1} s + e
+   * \f}
+   * where \f$D\f$ is the number of spatial dimensions, \f$d\f$ is the logical
+   * dimension of the neighbor from the element whose inbox this is, \f$s\f$
+   * is the side in the logical dimension \f$d\f$ with a vlaue of 1 for upper
```
Comment (kidder) [inline src/Evolution/DiscontinuousGalerkin/AtomicInboxBoundaryData.hpp]: value

File: src/Evolution/DiscontinuousGalerkin/AtomicInboxBoundaryData.hpp [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+   * for a specific `directional_id` that has been re-oriented using the
+   * `OrientationMap` to be put in the same block frame as the element that is
+   * receiving the data (i.e. that whose inbox this is being inserted into).
+   *
+   * The hash is computed as
+   * \f{align}{
+   * 2^D d + 2^{D-1} s + e
+   * \f}
+   * where \f$D\f$ is the number of spatial dimensions, \f$d\f$ is the logical
+   * dimension of the neighbor from the element whose inbox this is, \f$s\f$
```
Comment (kidder) [inline src/Evolution/DiscontinuousGalerkin/AtomicInboxBoundaryData.hpp]: dimension of the direction to the neighbor of the element

File: src/Evolution/DiscontinuousGalerkin/AtomicInboxBoundaryData.hpp [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+   * is the side in the logical dimension \f$d\f$ with a vlaue of 1 for upper
+   * and 0 for lower, and \f$e\f$ is the element segment hash. The element
```
Comment (kidder) [inline src/Evolution/DiscontinuousGalerkin/AtomicInboxBoundaryData.hpp]: is a hash of the index of the SegmentId%s of the neighbor's ElementId for the dimensions other than \f$d\f$.  In particular:  for \f$D = 1\f$; for  \f$D=1\f$, \f$e\f$ is 0 (1) if the SegmentId index along the face is even (odd); and for \f$D = 3\f$ \f$e\f$ is 0 (1, 2, 3) if the SegmentId indices along the face are both even (lower dim odd, higher dim odd, both dims odd).

File: src/Evolution/DiscontinuousGalerkin/AtomicInboxBoundaryData.hpp [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+   * segment hash is computed as the logical and of the SegmentID's index in
+   * that direction, left shifted by which direction on the face it is.
+   */
+  static size_t index(const DirectionalId<Dim>& directional_id);
+
+  std::array<Parallel::StaticSpscQueue<
```
Comment (kidder) [inline src/Evolution/DiscontinuousGalerkin/AtomicInboxBoundaryData.hpp:69]: can you add a code comment on where `20` comes from

**Review by kidder — CHANGES_REQUESTED** [review]
you can squash

**Comment by nilsdeppe** [comment]
Okay, rebased and squashed in the changes. Thanks for the reviews!

**Review by kidder — DISMISSED** [review]

**Comment by nilsdeppe** [comment]
Rebased since there were conflicts in the `ElementId` code

**Review by wthrowe — APPROVED** [review]

**Pull request merged**